packages/core-data/src/entities.ts

```diff
@@ -136,33 +188,126 @@ export const defaultEntities = [
 		label: __( 'Navigation Area' ),
 		key: 'name',
 		getTitle: ( record ) => record?.description,
-	},
-	{
+	} as const ),
+	defineEntity( {
 		label: __( 'Global Styles' ),
 		name: 'globalStyles',
 		kind: 'root',
 		baseURL: '/wp/v2/global-styles',
 		baseURLParams: { context: 'edit' },
 		plural: 'globalStylesVariations', // should be different than name
 		getTitle: ( record ) => record?.title?.rendered || record?.title,
-	},
-	{
+	} as const ),
+	defineEntity( {
 		label: __( 'Themes' ),
 		name: 'theme',
 		kind: 'root',
 		baseURL: '/wp/v2/themes',
 		baseURLParams: { context: 'edit' },
 		key: 'stylesheet',
-	},
-	{
+	} as const ),
+	defineEntity( {
 		label: __( 'Plugins' ),
 		name: 'plugin',
 		kind: 'root',
 		baseURL: '/wp/v2/plugins',
 		baseURLParams: { context: 'edit' },
 		key: 'plugin',
-	},
-];
+	} as const ),
+] as const;
+
+type Element = typeof defaultEntities[ number ];
+
+type DefaultEntityContext<
+	K extends string,
+	N extends string,
+	E = Extract< Element, { kind: K; name: N } >
+> = 'baseURLParams' extends keyof E
+	? 'context' extends keyof E[ 'baseURLParams' ]
+		? E[ 'baseURLParams' ][ 'context' ]
+		: 'view'
+	: 'view';
+
+type EntityKeyName<
+	K extends string,
+	N extends string,
+	E = Extract< Element, { kind: K; name: N } >
+> = 'key' extends keyof E ? E[ 'key' ] : 'id';
+
+type DeclaredEntity< K extends string, N extends string, RecordType > = {
+	kind: K;
+	name: N;
+	recordType: RecordType;
+	keyType: EntityKeyName< K, N > extends keyof RecordType
+		? RecordType[ EntityKeyName< K, N > ]
+		: never;
+	defaultContext: DefaultEntityContext< K, N >;
+};
+type APIEntity<
+	K extends string,
+	N extends string,
+	RecordType,
+	KeyName = 'id',
+	C = 'edit'
+> = {
+	kind: K;
+	name: N;
+	recordType: RecordType;
+	keyType: KeyName extends keyof RecordType ? RecordType[ KeyName ] : never;
+	defaultContext: C;
+};
+
+export type EntityTypeUnion< C extends Context = any > =
+	| DeclaredEntity< 'root', 'site', Settings< C > >
+	| DeclaredEntity< 'root', 'postType', Type< C > >
+	| DeclaredEntity< 'root', 'media', Attachment< C > >
+	| DeclaredEntity< 'root', 'taxonomy', Taxonomy< C > >
+	| DeclaredEntity< 'root', 'sidebar', Sidebar< C > >
+	| DeclaredEntity< 'root', 'widget', Widget< C > >
+	| DeclaredEntity< 'root', 'widgetType', WidgetType< C > >
+	| DeclaredEntity< 'root', 'user', User< C > >
+	| DeclaredEntity< 'root', 'comment', Comment< C > >
+	| DeclaredEntity< 'root', 'menu', NavMenu< C > >
+	| DeclaredEntity< 'root', 'menuItem', NavMenuItem< C > >
+	| DeclaredEntity< 'root', 'menuLocation', MenuLocation< C > >
+	| DeclaredEntity< 'root', 'navigationArea', NavigationArea< C > >
+	| DeclaredEntity< 'root', 'theme', Theme< C > >
+	| DeclaredEntity< 'root', 'plugin', Plugin< C > >
+	| APIEntity< 'postType', 'post', Post< C > >
+	| APIEntity< 'postType', 'page', Page< C > >
+	| APIEntity< 'postType', 'wp_template', WpTemplate< C > >
+	| APIEntity< 'postType', 'wp_template_part', WpTemplatePart< C > >;
```

Comment:
this is great. I love pushing the computation into the type system and when it matches the way we think about the actual code.